File: worker/test/jcs.test.ts
import { readFileSync, readdirSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { join } from "node:path";
import { describe, expect, it } from "vitest";
import { canonicalize } from "../src/jcs";

const corpusDir = fileURLToPath(
  new URL("../../conformance/canonical-json/", import.meta.url),
);

interface CorpusCase {
  name: string;
  input: unknown;
  expectedHex: string;
}

function loadCase(name: string): CorpusCase {
  const dir = join(corpusDir, name);
  // input.json is authoritative as raw bytes; parse those bytes as UTF-8 JSON.
  const inputBytes = readFileSync(join(dir, "input.json"));
  const input: unknown = JSON.parse(inputBytes.toString("utf-8"));
  // expected.hex is one line of lowercase hex plus a trailing LF (formatting).
  const expectedHex = readFileSync(join(dir, "expected.hex"), "utf-8").trim();
  expect(expectedHex).toMatch(/^[0-9a-f]+$/);
  return { name, input, expectedHex };
}

function toHex(bytes: Uint8Array): string {
  return Buffer.from(bytes).toString("hex");
}

const caseNames = readdirSync(corpusDir, { withFileTypes: true })
  .filter((entry) => entry.isDirectory())
  .map((entry) => entry.name)
  .sort();

const passGateCases = caseNames.filter((name) => !/^9\d\d-/.test(name));
const deliberateMismatchCases = caseNames.filter((name) =>
  /^9\d\d-/.test(name),
);

describe("corpus discovery", () => {
  it("finds at least 15 pass-gate cases (corpus is non-empty)", () => {
    expect(passGateCases.length).toBeGreaterThanOrEqual(15);
  });

  it("finds at least one deliberate-mismatch (9xx) case", () => {
    expect(deliberateMismatchCases.length).toBeGreaterThanOrEqual(1);
  });
});

describe("canonicalize matches shared vectors", () => {
  it.for(passGateCases)("%s", (name) => {
    const { input, expectedHex } = loadCase(name);
    expect(toHex(canonicalize(input))).toBe(expectedHex);
  });
});

describe("deliberate-mismatch (9xx) cases must NOT match correct output", () => {
  it.for(deliberateMismatchCases)("%s", (name) => {
    const { input, expectedHex } = loadCase(name);
    expect(toHex(canonicalize(input))).not.toBe(expectedHex);
  });
});

File: commonplace_log/test/jcs_test.exs
defmodule Commonplace.Log.JcsTest do
  @moduledoc """
  Conformance harness for the RFC 8785 canonicalizer, driven by the shared
  vector corpus in `conformance/canonical-json/`. The corpus is the arbiter:
  output must be byte-identical to `expected.hex`. Cases named `9xx-*` carry
  deliberately-wrong expected bytes and must MISMATCH (inverted assertion);
  they are excluded from the pass gate.
  """
  use ExUnit.Case, async: true

  @corpus_dir Path.expand("../../conformance/canonical-json", __DIR__)

  case_names =
    @corpus_dir
    |> File.ls!()
    |> Enum.filter(&File.dir?(Path.join(@corpus_dir, &1)))
    |> Enum.sort()

  {mismatch_cases, pass_cases} =
    Enum.split_with(case_names, &Regex.match?(~r/^9\d\d-/, &1))

  @pass_cases pass_cases
  @mismatch_cases mismatch_cases

  defp load_case(name) do
    dir = Path.join(@corpus_dir, name)

    # input.json is authoritative as raw bytes; parse those bytes here.
    input = dir |> Path.join("input.json") |> File.read!() |> Jason.decode!()

    # expected.hex is one line of lowercase hex; trailing LF is file formatting.
    expected_hex = dir |> Path.join("expected.hex") |> File.read!() |> String.trim()
    assert expected_hex =~ ~r/^[0-9a-f]+$/

    {input, expected_hex}
  end

  describe "corpus discovery (anti-vacuity)" do
    test "finds at least 18 pass-gate cases" do
      assert length(@pass_cases) >= 18
    end

    test "finds at least one deliberate-mismatch (9xx) case" do
      assert length(@mismatch_cases) >= 1
    end
  end

  describe "canonicalize matches shared vectors" do
    for name <- pass_cases do
      @case_name name
      test @case_name do
        {input, expected_hex} = load_case(@case_name)
        out = Commonplace.Log.Jcs.canonicalize(input)
        assert is_binary(out)
        assert Base.encode16(out, case: :lower) == expected_hex
      end
    end
  end

  describe "deliberate-mismatch (9xx) cases must NOT match correct output" do
    for name <- mismatch_cases do
      @case_name name
      test @case_name do
        {input, expected_hex} = load_case(@case_name)
        out = Commonplace.Log.Jcs.canonicalize(input)
        refute Base.encode16(out, case: :lower) == expected_hex
      end
    end
  end
end
